[Feature] Task card component

petertgiles opened the pull request:
🤖 Resolves #11075 

## 👋 Introduction

This branch adds a new task card component.

## 🕵️ Details

Add any additional details that could assist with reviewing or testing this PR.

## 🧪 Testing

1. ` pnpm run storybook:design-system`
2.  `localhost:6006/?path=/story/ui-src-components-taskcard--default`

- [ ] AC met
- [ ] light mode / dark mode
- [ ] standard mode / mobile mode

## 📸 Screenshot

![image](https://github.com/user-attachments/assets/cc3428f4-9a52-4c30-aa72-5f515026139f)

petertgiles commented on packages/ui/src/components/TaskCard/TaskCard.tsx:
```diff
@@ -0,0 +1,135 @@
+  icon?: IconType;
+  title: ReactNode;
+  headingColor?: CardColor;
+  link?: ReactElement<LinkProps>;
```
I noticed that Storybook flags an a11y issue on colour contrast of the link - but only when you switch modes.  I'm not sure if that's legitimate?  Also, I don't want to create a custom style exception for the links here.  That always causes pain when the standard styles move.  I'm not sure what to do here.

petertgiles replied:
The standard link colour is <color>.darker while the design uses <color>.darkest.  The standard design reserves the darkest colour variant for hover.

petertgiles replied:
Axe jest doesn't seem to mind.  :shrug:

yonikid15 replied:
Ya I'm also dealing with a color contrast issue on my component, which I'm not sure is legitimate either. Axe jest doesn't mind, as well as axe dev tools.

I checked the first elements contrast ratio using [WebAIM](https://webaim.org/resources/contrastchecker/) and it passed all the checks :shrug

![image](https://github.com/user-attachments/assets/eadeacbd-1fb9-41dc-ad63-29fb30e5c129)

yonikid15 replied:
Ya I'm also dealing with a color contrast issue on my component, which I'm not sure is legitimate either. Axe jest doesn't mind, as well as axe dev tools.

I checked the first elements contrast ratio using [WebAIM](https://webaim.org/resources/contrastchecker/) and it passed all the checks :shrug

![image](https://github.com/user-attachments/assets/eadeacbd-1fb9-41dc-ad63-29fb30e5c129)

yonikid15 reviewed (changes requested):
Nice work Peter, it's almost there!

This needs a bit of tweaking to match the mobile view: centre the heading and link within the parent div, and maybe .25 more padding (but that's up to @JoshBeveridge).

There isn't a design with the link in mobile design we got so you'll to ask Josh what the spacing should be.

petertgiles commented:
> This needs a bit of tweaking to match the mobile view: centre the heading and link within the parent div, and maybe .25 more padding

Good catch!  I've updated it.
![image](https://github.com/user-attachments/assets/6799f1d1-1dce-4e5e-9ce0-190e52a28fee)

yonikid15 commented on packages/ui/src/components/TaskCard/TaskCard.tsx:
```diff
@@ -0,0 +1,138 @@
+
+  // prepare heading text element
+  const HeadingTextElement = headingAs;
+  const commonHeadingTextStyles = { ...headingStyles["h4"] };
```
We should change the "h4" in `...headingStyles["h4"]` to the prop value `headingAs`.

petertgiles replied:
I'm not sure about this one.  I actually do want it to have all the styles of an h4.  
![image](https://github.com/user-attachments/assets/ca2d7a54-922e-4b05-afa6-0cd4290d4d6c)

Notice that semantic level is still headingAs.
![image](https://github.com/user-attachments/assets/69dcb6a9-f039-4cf7-99cc-9edfed3fbeb0)

yonikid15 replied:
Oh actually you can use the <Heading /> component instead. That way you don't need to import the styles and delete the margin. It would look something like:

```tsx
<Heading
  level={headingAs}
  size="h4"
  data-h2-margin="base(0)"
  data-h2-text-align="base(center) p-tablet(left)"
>
  {title}
</Heading>
```

petertgiles replied:
Great idaea!
[use Heading component](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/a1c5b895e83a0bd9c657f91e00eaecc256f0ce93)

yonikid15 commented on packages/ui/src/components/TaskCard/TaskCard.tsx:
```diff
@@ -0,0 +1,138 @@
+  children,
+}: TaskCardProps) => {
+  // prepare link
+  const headingBarLink = rawLink
+    ? cloneElement(rawLink, {
+        color: headingColor,
+      })
+    : null;
```
This is an [old practice](https://react.dev/reference/react/cloneElement) of React. To achieve the same result maybe change the prop to `link?: { href: string; label: string}`. Then you can use the Link component here directly, which should prevent the link from visual inconsistencies.

```tsx
<Link mode="text" href={link?.href} color={headingColor}>
    {link?.label}
 </Link>
```

yonikid15 reviewed (changes requested):
My bad Peter, I found some small things that need to be updated

petertgiles commented on packages/ui/src/components/TaskCard/TaskCard.tsx:
```diff
@@ -0,0 +1,138 @@
+        data-h2-border-radius="base(rounded rounded 0 0)"
+        data-h2-border-bottom-width="base(1px)"
+        data-h2-border-bottom-style="base(solid)"
+        data-h2-padding="base(x1) p-tablet(x1 x1.5 x1 x1.5)"
```
Right, updated in the same way as the other PR.
[terse styles](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/e8c9653537f7a797ee03e7facaab01d405cb3dcd)

yonikid15 reviewed (approved):
Excellent work :+1: Looks ready for a last design review from @JoshBeveridge then it's ready to merge

JoshBeveridge reviewed (changes requested):
Awesome! Sorry again for taking so long 😭

* If a link is present, it should use the horizontal space it needs to stay on a single line (see first screenshot)
* This might be a can of worms if you're using the `<Heading />` component, but the way the heading scales with longer text is pretty wonky (see first screenshot); the icon should always remain the same size and should have a consistent stroke width as determined by the heading, while the alignment of the icon should always remain vertically center with the first line of text
* The border separating the title bar from the card content is missing its color value (should be `color.darkest`)
* We can probably increase the gap between the title and the link on mobile devices to `x1`
* The interior padding of the card content should match the title bar (see second screenshot), but be careful here, because items inside the card based on the design require a divider that spans the full width - it might be easier to apply the padding to the items individually in this case (see third screenshot)

<details>
  <summary>Link span and icon size/alignment</summary>
  <img src="https://github.com/user-attachments/assets/e67fd289-354c-4517-b77e-58de15139285" />
</details> 

<details>
  <summary>Item padding</summary>
  <img src="https://github.com/user-attachments/assets/556b6ea5-6c50-4852-9339-e62b9a55a90c" />
</details> 

<details>
  <summary>Items requiring full-width borders</summary>
  <img src="https://github.com/user-attachments/assets/6047dd1a-344c-4ade-9799-5bce80704821" />
</details>

petertgiles commented:
Thanks for the review!

>    If a link is present, it should use the horizontal space it needs to stay on a single line (see first screenshot)

Updated in [don't wrap links](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/617646020dd573e2c44eef0f56eda6f84540f47c)

>    This might be a can of worms if you're using the <Heading /> component, but the way the heading scales with longer text is pretty wonky (see first screenshot); the icon should always remain the same size and should have a consistent stroke width as determined by the heading, while the alignment of the icon should always remain vertically center with the first line of text

OK, this customization won't be possible with the existing header component.  I've switched back to a customized header and included importing the icon styling from that component.  [switch back to custom heading](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/bc3f4625d11b9830885c6c463f4f663e99311533)

>    The border separating the title bar from the card content is missing its color value (should be color.darkest)

Good catch! [fix border colour](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/91c18dee65698cb2303b713592ed23d461091114)

>    We can probably increase the gap between the title and the link on mobile devices to x1

Update in [increase mobile heading gap](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/02897bf7b515d7de7e6688338023b5aaa188c3ea)

>    The interior padding of the card content should match the title bar (see second screenshot), but be careful here, because items inside the card based on the design require a divider that spans the full width - it might be easier to apply the padding to the items individually in this case (see third screenshot)

This was actually a bug in the story so I fixed that in [fix storybook story](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/68a57f62837d289303624dfb84b731c719871ba3).  However, this will probably be common enough that I added a lightweight wrapper for easy use.  [TaskCardItem component](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/a56a4d861bb68c17a201d271e7aae258c7b1dfdd)

yonikid15 commented on packages/ui/src/components/TaskCard/TaskCardItem.tsx:
```diff
@@ -0,0 +1,18 @@
+  children?: ReactNode;
+}
+
+const TaskCardItem = ({ children }: TaskCardItemProps) => {
+  return (
+    <div
+      data-h2-padding="base(x1) p-tablet(x1 x1.5)"
+      data-h2-border-bottom="base:selectors[:not(:last-child)](1px solid gray.lighter)"
+    >
+      {children}
+    </div>
+  );
+};
```
I don't think this is what Josh meant, cause this component is only a card. The heading padding should match the interior padding, so you only needed to add the padding to the `<div>{children}</div>`. However, looking at the screenshot of the #10785 Dashboard view, this would keep things consistent so I like it :+1: !

Sorry to be nit picky, but could we make this and TaskCard into a compound component instead? Similar to the  [DescriptionList](https://github.com/GCTC-NTGC/gc-digital-talent/blob/11075-task-card-component/packages/ui/src/components/DescriptionList/DescriptionList.tsx#L1-L115). You just need to add this into the TaskCard, and rename this to Item, and rename the TaskCard to Root. Then we can use it similar to our other components: `<TaskCard.Root /><TaskCard.Item>{...}</TaskCard.Item></TaskCard.Root>`

petertgiles replied:
I like consistency.  If we end up using this card for non-list content we might have to undo this though.
[refactor to compound component](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/7b4c5222782220a76328aba072e307297ad58d2c)

yonikid15 reviewed (changes requested):
Just two small changes that I suggested, and then this looks good to go!

petertgiles commented on packages/ui/src/components/TaskCard/TaskCard.tsx:
```diff
@@ -0,0 +1,141 @@
+        {/* wrapper */}
+        <div {...wrapperStyleMap[headingColor]} data-h2-flex-grow="base(2)">
+          <CustomHeading
+            {...headingStyles["h4"]}
```
Yeah, I guess it is possible.  It makes me cringe that it has to customized so much - at what point is it still worth trying to use the existing component?
```tsx
<Heading
  level={headingAs}
  size="h4"
  data-h2-margin="base(0)"
  data-h2-text-align="base(center) p-tablet(left)"
  data-h2-text-wrap="base(balance)"
  {...(Icon && {
    "data-h2-display": "base(flex)",
    "data-h2-align-items": "base(start)",
    "data-h2-gap": "base(0 x.5)",
  })}
  Icon={icon}
>
  {title}
</Heading>
```

I guess I don't like it any _less_ this way.  :laughing: 
[use heading component again](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11161/commits/163ed1ac085179ed39ae3954fbe7402f05c2e12e)

yonikid15 replied:
Well that's true you were using the headingStyles anyways, so it's basically the same thing :laughing:

petertgiles commented:
Yikes, I forgot about that.  Yeah, one of the big reasons I have to do a custom header is that the icon is supposed to disappear at mobile widths.  I rolled back to my custom header but to make it a bit more palatable I factored it out to its own subcomponent with some extra commenting.

yonikid15 reviewed (approved):
:1st_place_medal:
